moshi-sealed/metadata-reflect/src/main/kotlin/dev/zacsweers/moshix/sealed/reflect/MetadataMoshiSealedJsonAdapterFactory.kt

```diff
@@ -0,0 +1,152 @@
+package dev.zacsweers.moshix.sealed.reflect
+
+import com.squareup.moshi.JsonAdapter
+import com.squareup.moshi.JsonClass
+import com.squareup.moshi.Moshi
+import com.squareup.moshi.Types
+import com.squareup.moshi.adapters.PolymorphicJsonAdapterFactory
+import dev.zacsweers.moshix.sealed.annotations.DefaultNull
+import dev.zacsweers.moshix.sealed.annotations.DefaultObject
+import dev.zacsweers.moshix.sealed.annotations.TypeLabel
+import dev.zacsweers.moshix.sealed.runtime.internal.ObjectJsonAdapter
+import kotlinx.metadata.ClassName
+import kotlinx.metadata.Flag
+import kotlinx.metadata.KmClass
+import kotlinx.metadata.jvm.KotlinClassHeader
+import kotlinx.metadata.jvm.KotlinClassMetadata
+import java.lang.reflect.Type
+
+/** Classes annotated with this are eligible for this adapter. */
+private val KOTLIN_METADATA = Metadata::class.java
+
+private val UNSET = Any()
+
+public class MetadataMoshiSealedJsonAdapterFactory : JsonAdapter.Factory {
+  override fun create(type: Type,
+                      annotations: MutableSet<out Annotation>,
+                      moshi: Moshi): JsonAdapter<*>? {
+    if (annotations.isNotEmpty()) {
+      return null
+    }
+    val rawType = Types.getRawType(type)
+    if (!rawType.isAnnotationPresent(KOTLIN_METADATA)) return null
+    val kmClass = rawType.header()?.toKmClass() ?: return null
+
+    rawType.getAnnotation(JsonClass::class.java)?.let { jsonClass ->
+      val generator = jsonClass.generator
+      if (!generator.startsWith("sealed:")) {
+        return null
+      }
+      val typeLabel = generator.removePrefix("sealed:")
+      if (!Flag.IS_SEALED(kmClass.flags)) {
+        return null
```

Comment:
Not for this PR, but wonder if we should log some sort of warning here. I suppose it's possibly another generator is coming in but seems unlikely

Reply:
Such warning would probably help people to spot human mistake when everything is set up correctly but `sealed` keyword is missing on the class. That would require making the class abstract, though, so yeah it's unlikely too.